src/main/java/play/dev/filewatch/SourceModificationWatch.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package play.dev.filewatch;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+public final class SourceModificationWatch {
+
+  private SourceModificationWatch() {}
+
+  private static Set<File> listFiles(Supplier<Iterable<File>> sourcesFinder) {
+    return StreamSupport.stream(sourcesFinder.get().spliterator(), false)
+        .collect(Collectors.toSet());
+  }
+
+  private static long findLastModifiedTime(File file) {
+    try {
+      return Files.getLastModifiedTime(file.toPath()).toInstant().toEpochMilli();
+    } catch (IOException e) {
+      return 0L;
+    }
+  }
+
+  private static long findLastModifiedTime(Set<File> files) {
+    if (!files.isEmpty()) {
+      return files.stream()
+          .max(Comparator.comparingLong(SourceModificationWatch::findLastModifiedTime))
+          .map(SourceModificationWatch::findLastModifiedTime)
+          .get();
+    } else {
+      return 0L;
+    }
+  }
+
+  public static class Result {
+    private final boolean triggered;
+
+    private final WatchState state;
+
+    public boolean isTriggered() {
+      return triggered;
+    }
+
+    public WatchState getState() {
+      return state;
+    }
+
+    public Result(boolean triggered, WatchState state) {
+      this.triggered = triggered;
+      this.state = state;
+    }
+  }
+
+  public static Result watch(
+      final Supplier<Iterable<File>> sourcesFinder,
+      final int pollDelayMillis,
+      WatchState state,
+      final Supplier<Boolean> terminationCondition) {
+    while (true) {
+      var filesToWatch = listFiles(sourcesFinder);
+
+      var sourceFilesPath =
+          filesToWatch.stream()
+              .map(
+                  file -> {
+                    try {
+                      return file.getCanonicalPath();
+                    } catch (IOException e) {
+                      return null;
+                    }
+                  })
+              .filter(Objects::nonNull)
+              .collect(Collectors.toSet());
+      var lastModifiedTime = findLastModifiedTime(filesToWatch);
+
+      var sourcesModified =
+          lastModifiedTime > state.getLastCallbackCallTime()
+              || !state.getPreviousFiles().equals(sourceFilesPath);
+
+      var triggered = sourcesModified ? false : state.isAwaitingQuietPeriod();
+      var newCallbackCallTime =
+          sourcesModified ? System.currentTimeMillis() : state.getLastCallbackCallTime();
```

Comment:
I used elvis operator here to keep this in one line instead of using an `if` afterwards. In your code you always call `isAwaitingQuietPeriod` and `getLastCallbackCallTime`, even if it will be overriden in the if.
For this code it does not matter, but what if `isAwaitingQuietPeriod` and `getLastCallbackCallTime` would do some heavy processing?
I prefer to call methods only if necessary.